src/Proto.Cluster.Identity/IdentityStorageWorker.cs

```diff
@@ -32,51 +32,74 @@ public IdentityStorageWorker(IdentityStorageLookup storageLookup)
             _storage = storageLookup.Storage;
         }
 
-        public async Task ReceiveAsync(IContext context)
+        public Task ReceiveAsync(IContext context)
         {
-            if (context.Message is not GetPid msg) return;
+            if (context.Message is not GetPid msg) return Task.CompletedTask;
 
             if (context.Sender == null)
             {
                 _logger.LogCritical("No sender in GetPid request");
-                return;
+                return Task.CompletedTask;
             }
 
             if (_cluster.PidCache.TryGet(msg.ClusterIdentity, out var existing))
             {
-                Respond(existing);
-                return;
+                context.Respond(new PidResult
+                    {
+                        Pid = existing
+                    }
+                );
+                return Task.CompletedTask;
             }
 
-            try
+            if (_inProgress.TryGetValue(msg.ClusterIdentity, out var pidList))
             {
-                var pid = await GetWithGlobalLock(context.Sender!, msg.ClusterIdentity, context.CancellationToken);
-                if (pid != null) _cluster.PidCache.TryAdd(msg.ClusterIdentity, pid);
-
-                Respond(pid);
-            }
-            catch (Exception x)
-            {
-                _logger.LogError(x, "Identity worker crashed {Id}", context.Self!.ToShortString());
-                throw;
+                pidList.Add(context.Sender);
+                return Task.CompletedTask;
             }
 
+            _inProgress.Add(msg.ClusterIdentity, new List<PID>());
 
-            void Respond(PID? pid)
-            {
-                context.Respond(new PidResult
+            context.ReenterAfter(GetWithGlobalLock(context.Sender!, msg.ClusterIdentity, context.CancellationToken), task => {
```

Comment:
I like this 👍🏻 , let's consider if it can be made into some form of pattern/abstraction also.
IIRC we have(had?) a similar case in PartitionIdentityActor.

TaskCompleter<TKey,TValue,TSubscriber>  _taskCompleter = new();


_taskCompleter.Add (key,  onMissing: key => ...., onCompleted: (value, subscriberArr) => foreach(var s in subscriberArr) ..... )

something in that direction maybe?

Reply:
Not a bad idea. Perhaps we can make it more efficient in the base case as well, eliminating the list allocation?